ceph/ceph
# mgr/dashboard: clean-up python unit tests

## Dashboard Python Unit Test Refactoring

Now Python unit tests don't require Ceph to built before in order to run. All dependencies have been removed or mocked.

Unit-testing:
- Add conftest.py to mock imported modules (rados, rbd, cephfs), and mock
also rados Error and OSError Exceptions.
- Fix pylint issues.
- Fix RST issues (HACKING.rst)
- Refactor JwtManager check to avoid code duplication (found by pylint when --jobs=1).
- Fix issue with DashboardException.code property, using abs() on potentially None attribute.
- Disables Doctests in services/rbd.py that are actually integration tests (they check the value of rbd.RBD_FEATURES_NAME_MAPPING). Ideally, these kind of tests should be explicitly executed in an integration testing stage, rather that unit-testing. Disabled tests have been prepended with @DISABLEDOCTEST token.

Refactor tox.ini:
- Remove requirements-{py27,py3}.txt, as python release dependant
packages can be handled with PEP 508 syntax.
- Remove develepment dependencies from requirements.
- Move pycodestyle settings to separate section.
- Add flake8 check
- Add RST linter.
- Add instafail plugin to report failures immediately
- Add timeout plugin to limit max run time (sometimes test_tasks hangs)
- Remove unused dependencies (lru_cache, pluggy)

Docs:
 - Update HACKING.rst

Fixes: https://tracker.ceph.com/issues/40487
Fixes: https://tracker.ceph.com/issues/41152
Signed-off-by: Ernesto Puerta <epuertat@redhat.com>

---
